feat: Support l10n extensions

diegoveloper opened the pull request:
Support l10n extensions

[ ] Documentation update
[ ] Bug fix
[ ] New rule
[ ] Changes an existing rule
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

- I didn't include tests because:
   - Current tests are not working on my local.
 
- Documentation is not clear about how to use `check-unused-l10n`. https://dartcodemetrics.dev/docs/cli/check-unused-l10n
The -p param would be used to find files with that pattern too, is not just for Class names.


### What changes did you make? (Give an overview)
- The changes I made were on `UnusedL10nVisitor` class, now we support unused code called from extensions, the only requirements to find unused Strings on extension are:
    -  The file name where the `Localization` is implemented should contain the name of the getter for the extension. Eg:

```dart
filename: class_with_l10n.dart

class DemoLocalizations {
  DemoLocalizations(this.localeName);

 final String localeName;
  
  String get hello {
    return Intl.message(
      'Hello World',
      name: 'hello,
      locale: localeName,
    );
  }
  ...
}
```

extension:
```dart
filename: anyname.dart

extension MyLocalizationExtension on BuildContext {
  DemoLocalizations get l10n => DemoLocalizations.of(this);
}
```
`l10n` is the name that we are going to search, then the usage will be :

```dart
flutter pub run dart_code_metrics:metrics check-unused-l10n lib -p l10n
```
Where `l10n` will find the class inside `class_with_l10n.dart` and the getter `l10n` from the extension  `MyLocalizationExtension`.

NOTE: I'm sure that you could improve this (maybe another param?).


### Is there anything you'd like reviewers to focus on?
- Yes, help me to improve the documentation and adding unit tests or any other change that this PR requires in order to improve the detection of unused l10n strings (extensions or any other class).

dkrutskikh commented:
@diegoveloper first of all, thanks for contribute! Please update changelog

diegoveloper commented:
> @diegoveloper first of all, thanks for contribute! Please update changelog

done :)

incendial commented:
> I didn't include tests because: Current tests are not working on my local.

Could you share more details about the problem? What's not working? And have you run your changes on a real code?

> The -p param would be used to find files with that pattern too, is not just for Class names.

Hm, but right now it really only checks the class names and nothing else. Do you have an example where it's not true?

> flutter pub run dart_code_metrics:metrics check-unused-l10n lib -p l10n

I think that it kinda violates the current approach with class names search. Instead we could change the code to correctly work with `flutter pub run dart_code_metrics:metrics check-unused-l10n lib -p DemoLocalizations`. What do you think?

incendial commented:
@diegoveloper and yeah, so sorry for not answering, I'm kinda overloaded right now, trying to deal with it. Thank you very much for opening this PR!

diegoveloper commented:
> > I didn't include tests because: Current tests are not working on my local.
> 
> Could you share more details about the problem? What's not working? And have you run your changes on a real code?
> 

Yes, my changes are working on my local code, but the tests were not working even before my changes:
These 3 tests are failing:

```dart
01:58 +313 -1: test/src/cli/commands/check_unused_files_command_test.dart: CheckUnusedFilesCommand should have correct usage [E]    
  Expected: 'Check unused *.dart files.\n'
  ````

```dart
01:59 +316 -2: test/src/cli/commands/analyze_command_test.dart: AnalyzeCommand should have correct usage [E]                        
  Expected: 'Collect code metrics, rules and anti-patterns violations.\n'
```
```dart
01:59 +320 -3: test/src/cli/commands/check_unused_l10n_command_test.dart: CheckUnusedL10nCommand should have correct usage [E]      
  Expected: 'Check unused localization in *.dart files.\n'
```

> > The -p param would be used to find files with that pattern too, is not just for Class names.
> 
> Hm, but right now it really only checks the class names and nothing else. Do you have an example where it's not true?

Yeah, I was trying different combinations, for example if we use: 

```dart
flutter pub run dart_code_metrics:metrics check-unused-l10n lib -p l10n
```
And it doesn't find any class with that name, it search by file name like this :

```dart
filename: class_with_l10n.dart

class DemoLocalizations {
  DemoLocalizations(this.localeName);

 final String localeName;
  
  String get hello {
    return Intl.message(
      'Hello World',
      name: 'hello,
      locale: localeName,
    );
  }
  ...
}
```
Then I have my extension in another file: 

```dart
filename: anyname.dart

extension MyLocalizationExtension on BuildContext {
  DemoLocalizations get l10n => DemoLocalizations.of(this);
}
```

And I reuse the pattern "l10n" to find the correct getter.

> 
> > flutter pub run dart_code_metrics:metrics check-unused-l10n lib -p l10n
> 
> I think that it kinda violates the current approach with class names search. Instead we could change the code to correctly work with `flutter pub run dart_code_metrics:metrics check-unused-l10n lib -p DemoLocalizations`. What do you think?

Yeah, it's kinda confusing, it would be better to create new params (for files and for classnames + extensions). 
The changes I sent are working on my project but as you said (with extensions), it kinda violates the approach.

incendial commented:
> Yes, my changes are working on my local code, but the tests were not working even before my changes:

Interesting, I can't reproduce it by myself. Could you share how you run them and your dart sdk and os?

> And it doesn't find any class with that name, it search by file name like this :

The new code that you've added searches for getters that named as -p (in your case `l10n`) and then tracks all usages of the class of returned value. That's why you might have this idea, but, no, it really searches for classes, not files.

incendial commented on lib/src/analyzers/unused_l10n_analyzer/unused_l10n_visitor.dart:
```diff
@@ -117,4 +127,22 @@ class UnusedL10nVisitor extends RecursiveAstVisitor<void> {
+    final staticElement =
+        target.staticElement?.enclosingElement as ExtensionElement;
+    for (final element in staticElement.accessors) {
+      if (_classPattern.hasMatch(element.name) && element.isGetter) {
```
This is the place: you basically check for a getter and match. Instead, I'd suggest to check the return value of a getter for a patter match. That would fit into the whole command idea.

diegoveloper replied:
Yeah, the problem is my file (that contains the extension) wasn't included when I was using -p `DemoLocalizations`.

incendial replied:
With the current code?

diegoveloper replied:
I think you are right, not sure why I was getting different results, I'll check my code again and I'll send some changes :D thanks

diegoveloper replied:
@incendial I sent my changes , can you check please? do you have any hints about the test? I'm using `dart test`.

diegoveloper commented:
> > Yes, my changes are working on my local code, but the tests were not working even before my changes:
> 
> Interesting, I can't reproduce it by myself. Could you share how you run them and your dart sdk and os?
> 

I just used:  `dart test` 

> > And it doesn't find any class with that name, it search by file name like this :
> 
> The new code that you've added searches for getters that named as -p (in your case `l10n`) and then tracks all usages of the class of returned value. That's why you might have this idea, but, no, it really searches for classes, not files.

When I used `DemoLocalizations` in `-p` , the file : 

```
filename: anyname.dart

extension MyLocalizationExtension on BuildContext {
  DemoLocalizations get l10n => DemoLocalizations.of(this);
}
```

Was not checked, therefore I couldn't compare the return type from the getters.

incendial commented:
@diegoveloper the changes looks great, but I have no idea, why the tests are failing 😞 Do you have MacOS?

diegoveloper commented:
> @diegoveloper the changes look great, but I have no idea, why the tests are failing 😞 Do you have MacOS?

Yes, 

```
[✓] Flutter (Channel stable, 2.8.1, on macOS 12.0.1 21A559 darwin-arm, locale en-PE)
[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2020.3)
[✓] VS Code (version 1.63.2)
[✓] Connected device (2 available)
```

dkrutskikh commented:
> Yes, my changes are working on my local code, but the tests were not working even before my changes:

@diegoveloper can you provide full local path to source code?

diegoveloper commented:
> > Yes, my changes are working on my local code, but the tests were not working even before my changes:
> 
> @diegoveloper can you provide full local path to source code?

@dkrutskikh 

```
/Users/diegoveloper/Development/git/dart_code_metrics
```

When I'm on the root path I just run:

```
dart test
```

diegoveloper commented:
@incendial @dkrutskikh This is the full log of the error :

```
diegoveloper@diegovelopers-MacBook-Pro dart_code_metrics % dart test
03:21 +322 -2: test/src/cli/commands/check_unused_files_command_test.dart: CheckUnusedFilesCommand should have correct usage [E]                                                                                                                  
  Expected: 'Check unused *.dart files.\n'
              '\n'
              'Usage: metrics check-unused-files [arguments] <directories>\n'
              '-h, --help                                        Print this usage information.\n'
              '\n'
              '\n'
              '-r, --reporter=<console>                          The format of the output of the analysis.\n'
              '                                                  [console (default), json]\n'
              '\n'
              '\n'
              '    --root-folder=<./>                            Root folder.\n'
              '                                                  (defaults to current directory)\n'
              '    --sdk-path=<directory-path>                   Dart SDK directory path. Should be provided only when you run the application as compiled executable(https://dart.dev/tools/dart-compile#exe) and automatic Dart SDK path detection fails.\n'
              '    --exclude=<{/**.g.dart,/**.template.dart}>    File paths in Glob syntax to be exclude.\n'
              '                                                  (defaults to "{/**.g.dart,/**.template.dart}")\n'
              '\n'
              '\n'
              '    --[no-]fatal-unused                           Treat find unused file as fatal.\n'
              '\n'
              'Run "metrics help" to see global options.'
    Actual: 'Check unused *.dart files.\n'
              '\n'
              'Usage: metrics check-unused-files [arguments] <directories>\n'
              '-h, --help                                        Print this usage information.\n'
              '\n'
              '\n'
              '-r, --reporter=<console>                          The format of the output of the analysis.\n'
              '                                                  [console (default), json]\n'
              '\n'
              '\n'
              '    --root-folder=<./>                            Root folder.\n'
              '                                                  (defaults to "/Users/diegoveloper/Development/git/tonal/dart_code_metrics")\n'
              '    --sdk-path=<directory-path>                   Dart SDK directory path. Should be provided only when you run the application as compiled executable(https://dart.dev/tools/dart-compile#exe) and automatic Dart SDK path detection fails.\n'
              '    --exclude=<{/**.g.dart,/**.template.dart}>    File paths in Glob syntax to be exclude.\n'
              '                                                  (defaults to "{/**.g.dart,/**.template.dart}")\n'
              '\n'
              '\n'
              '    --[no-]fatal-unused                           Treat find unused file as fatal.\n'
              '\n'
              'Run "metrics help" to see global options.'
     Which: is different.
            Expected: ... faults to current di ...
              Actual: ... faults to "/Users/di ...
                                    ^
             Differ at offset 477
  
  package:test_api                                                 expect
  test/src/cli/commands/check_unused_files_command_test.dart 52:7  main.<fn>.<fn>
  
03:21 +324 -2: test/src/cli/commands/analyze_command_test.dart: AnalyzeCommand should have correct usage [E]                                                                                                                                      
  Expected: 'Collect code metrics, rules and anti-patterns violations.\n'
              '\n'
              'Usage: metrics analyze [arguments] <directories>\n'
              '-h, --help                                        Print this usage information.\n'
              '\n'
              '\n'
              '-r, --reporter=<console>                          The format of the output of the analysis.\n'
              '                                                  [console (default), console-verbose, codeclimate, github, gitlab, html, json]\n'
              '-o, --output-directory=<OUTPUT>                   Write HTML output to OUTPUT.\n'
              '                                                  (defaults to "metrics")\n'
              '\n'
              '\n'
              '    --cyclomatic-complexity=<20>                  Cyclomatic Complexity threshold.\n'
              '    --halstead-volume=<150>                       Halstead Volume threshold.\n'
              '    --lines-of-code=<100>                         Lines of Code threshold.\n'
              '    --maximum-nesting-level=<5>                   Maximum Nesting Level threshold.\n'
              '    --number-of-methods=<10>                      Number of Methods threshold.\n'
              '    --number-of-parameters=<4>                    Number of Parameters threshold.\n'
              '    --source-lines-of-code=<50>                   Source lines of Code threshold.\n'
              '    --weight-of-class=<0.33>                      Weight Of a Class threshold.\n'
              '    --maintainability-index=<50>                  Maintainability Index threshold.\n'
              '    --technical-debt=<0>                          Technical Debt threshold.\n'
              '\n'
              '\n'
              '    --root-folder=<./>                            Root folder.\n'
              '                                                  (defaults to current directory)\n'
              '    --sdk-path=<directory-path>                   Dart SDK directory path. Should be provided only when you run the application as compiled executable(https://dart.dev/tools/dart-compile#exe) and automatic Dart SDK path detection fails.\n'
              '    --exclude=<{/**.g.dart,/**.template.dart}>    File paths in Glob syntax to be exclude.\n'
              '                                                  (defaults to "{/**.g.dart,/**.template.dart}")\n'
              '\n'
              '\n'
              '    --set-exit-on-violation-level=<warning>       Set exit code 2 if code violations same or higher level than selected are detected.\n'
              '                                                  [noted, warning, alarm]\n'
              '    --[no-]fatal-style                            Treat style level issues as fatal.\n'
              '    --[no-]fatal-performance                      Treat performance level issues as fatal.\n'
              '    --[no-]fatal-warnings                         Treat warning level issues as fatal.\n'
              '\n'
              'Run "metrics help" to see global options.'
    Actual: 'Collect code metrics, rules and anti-patterns violations.\n'
              '\n'
              'Usage: metrics analyze [arguments] <directories>\n'
              '-h, --help                                        Print this usage information.\n'
              '\n'
              '\n'
              '-r, --reporter=<console>                          The format of the output of the analysis.\n'
              '                                                  [console (default), console-verbose, codeclimate, github, gitlab, html, json]\n'
              '-o, --output-directory=<OUTPUT>                   Write HTML output to OUTPUT.\n'
              '                                                  (defaults to "metrics")\n'
              '\n'
              '\n'
              '    --cyclomatic-complexity=<20>                  Cyclomatic Complexity threshold.\n'
              '    --halstead-volume=<150>                       Halstead Volume threshold.\n'
              '    --lines-of-code=<100>                         Lines of Code threshold.\n'
              '    --maximum-nesting-level=<5>                   Maximum Nesting Level threshold.\n'
              '    --number-of-methods=<10>                      Number of Methods threshold.\n'
              '    --number-of-parameters=<4>                    Number of Parameters threshold.\n'
              '    --source-lines-of-code=<50>                   Source lines of Code threshold.\n'
              '    --weight-of-class=<0.33>                      Weight Of a Class threshold.\n'
              '    --maintainability-index=<50>                  Maintainability Index threshold.\n'
              '    --technical-debt=<0>                          Technical Debt threshold.\n'
              '\n'
              '\n'
              '    --root-folder=<./>                            Root folder.\n'
              '                                                  (defaults to "/Users/diegoveloper/Development/git/tonal/dart_code_metrics")\n'
              '    --sdk-path=<directory-path>                   Dart SDK directory path. Should be provided only when you run the application as compiled executable(https://dart.dev/tools/dart-compile#exe) and automatic Dart SDK path detection fails.\n'
              '    --exclude=<{/**.g.dart,/**.template.dart}>    File paths in Glob syntax to be exclude.\n'
              '                                                  (defaults to "{/**.g.dart,/**.template.dart}")\n'
              '\n'
              '\n'
              '    --set-exit-on-violation-level=<warning>       Set exit code 2 if code violations same or higher level than selected are detected.\n'
              '                                                  [noted, warning, alarm]\n'
              '    --[no-]fatal-style                            Treat style level issues as fatal.\n'
              '    --[no-]fatal-performance                      Treat performance level issues as fatal.\n'
              '    --[no-]fatal-warnings                         Treat warning level issues as fatal.\n'
              '\n'
              'Run "metrics help" to see global options.'
     Which: is different.
            Expected: ... faults to current di ...
              Actual: ... faults to "/Users/di ...
                                    ^
             Differ at offset 1517
  
  package:test_api                                      expect
  test/src/cli/commands/analyze_command_test.dart 69:7  main.<fn>.<fn>
  
03:22 +324 -3: test/src/cli/commands/check_unused_l10n_command_test.dart: CheckUnusedL10nCommand should have correct usage [E]                                                                                                                    
  Expected: 'Check unused localization in *.dart files.\n'
              '\n'
              'Usage: metrics check-unused-l10n [arguments] <directories>\n'
              '-h, --help                                        Print this usage information.\n'
              '\n'
              '\n'
              '-p, --class-pattern=<I18n$>                       The pattern to detect classes providing localization\n'
              '                                                  (defaults to "I18n$")\n'
              '\n'
              '\n'
              '-r, --reporter=<console>                          The format of the output of the analysis.\n'
              '                                                  [console (default), json]\n'
              '\n'
              '\n'
              '    --root-folder=<./>                            Root folder.\n'
              '                                                  (defaults to current directory)\n'
              '    --sdk-path=<directory-path>                   Dart SDK directory path. Should be provided only when you run the application as compiled executable(https://dart.dev/tools/dart-compile#exe) and automatic Dart SDK path detection fails.\n'
              '    --exclude=<{/**.g.dart,/**.template.dart}>    File paths in Glob syntax to be exclude.\n'
              '                                                  (defaults to "{/**.g.dart,/**.template.dart}")\n'
              '\n'
              '\n'
              '    --[no-]fatal-unused                           Treat find unused l10n as fatal.\n'
              '\n'
              'Run "metrics help" to see global options.'
    Actual: 'Check unused localization in *.dart files.\n'
              '\n'
              'Usage: metrics check-unused-l10n [arguments] <directories>\n'
              '-h, --help                                        Print this usage information.\n'
              '\n'
              '\n'
              '-p, --class-pattern=<I18n$>                       The pattern to detect classes providing localization\n'
              '                                                  (defaults to "I18n$")\n'
              '\n'
              '\n'
              '-r, --reporter=<console>                          The format of the output of the analysis.\n'
              '                                                  [console (default), json]\n'
              '\n'
              '\n'
              '    --root-folder=<./>                            Root folder.\n'
              '                                                  (defaults to "/Users/diegoveloper/Development/git/tonal/dart_code_metrics")\n'
              '    --sdk-path=<directory-path>                   Dart SDK directory path. Should be provided only when you run the application as compiled executable(https://dart.dev/tools/dart-compile#exe) and automatic Dart SDK path detection fails.\n'
              '    --exclude=<{/**.g.dart,/**.template.dart}>    File paths in Glob syntax to be exclude.\n'
              '                                                  (defaults to "{/**.g.dart,/**.template.dart}")\n'
              '\n'
              '\n'
              '    --[no-]fatal-unused                           Treat find unused l10n as fatal.\n'
              '\n'
              'Run "metrics help" to see global options.'
     Which: is different.
            Expected: ... faults to current di ...
              Actual: ... faults to "/Users/di ...
                                    ^
             Differ at offset 673
  
  package:test_api                                                expect
  test/src/cli/commands/check_unused_l10n_command_test.dart 56:7  main.<fn>.<fn>
  
03:33 +329 -3: Some tests failed.                                                                                                                                                                                                                 

Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
diegoveloper@diegovelopers-MacBook-Pro dart_code_metrics % 
```

Hmm so the problem is related to the path right? should I consider anything before running the tests? 

```
   Expected: ... faults to current di ...
              Actual: ... faults to "/Users/di ...
              ```

dkrutskikh commented:
@diegoveloper thank you, I found the root of the problem. I think we will fix it soon.

dkrutskikh commented:
@diegoveloper we had merge fix in master branch. Can you rebase on head or merge master in your branch?

diegoveloper commented:
> @diegoveloper we had merge fix in master branch. Can you rebase on head or merge master in your branch?

Great, I updated my PR, now all the tests passed 😄

incendial commented:
> > @diegoveloper we had merge fix in master branch. Can you rebase on head or merge master in your branch?
> 
> 
> 
> Great, I updated my PR, now all the tests passed 😄 

Could you add new test cases to cover extensions support?

diegoveloper commented:
> > > @diegoveloper we had merge fix in master branch. Can you rebase on head or merge master in your branch?
> > 
> > 
> > Great, I updated my PR, now all the tests passed 😄
> 
> Could you add new test cases to cover extensions support?

Hey @incendial , I added a new unit test to cover extensions, can you check again to see if the coverage is not going down ?

incendial commented:
@diegoveloper the test is great, thank you! And thank you so much for the contribution! 🔥🚀

incendial commented:
We'll release 4.9.0 in a few days, when some new rules will be finished.
